File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
    static int n, l, r;
    static int[][] a;
    static boolean[][] visited;
    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};
    static int sum = 0, ret = 0;
    static boolean flag = false;
    static ArrayList<Pair> tmp = new ArrayList<>();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        l = sc.nextInt();
        r = sc.nextInt();
        a = new int[n][n];
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                a[i][j] = sc.nextInt();
            }
        }
        while(true){
            flag = false;
            visited = new boolean[n][n];
            for(int i = 0; i < n; i++){
                for(int j = 0; j < n; j++){
                    if(visited[i][j]) continue;
                    sum = a[i][j];
                    tmp.add(new Pair(i, j));
                    dfs(i, j);
                    int cnt = tmp.size();
                    if(cnt > 1) flag = true;
                    for(Pair p : tmp){
                        a[p.y][p.x] = sum / cnt;
                    }
                    tmp.clear();
                }
            }
            if(!flag) break;
            ret++;
        }
        System.out.println(ret);
    }

    private static void dfs(int y, int x){
        visited[y][x] = true;
        for(int i = 0; i < 4; i++){
            int ny = y + dy[i];
            int nx = x + dx[i];
            if(ny < 0 || nx < 0 || ny >= n || nx >= n) continue;
            if(visited[ny][nx]) continue;
            int diff = Math.abs(a[y][x] - a[ny][nx]);
            if(diff >= l && diff <= r){
                sum += a[ny][nx];
                tmp.add(new Pair(ny, nx));
                dfs(ny, nx);
            }
        }
    }
}

class Pair{
    int y;
    int x;
    Pair(int y, int x){
        this.y = y;
        this.x = x;
    }
}

File: 백준/Gold/16234. 인구 이동/인구 이동.cc
#include<bits/stdc++.h>
using namespace std;

int n, l, r, ret = 0;
int a[54][54], visited[54][54];
int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, 1, 0, -1};
vector<pair<int, int>> v;

void dfs(int y, int x){
	visited[y][x] = 1;
	v.push_back({y, x});
	for(int i = 0; i < 4; i++){
		int ny = y + dy[i];
		int nx = x + dx[i];
		if(ny < 0 || nx < 0 || ny >= n || nx >= n || visited[ny][nx] == 1) continue;
		if(abs(a[y][x] - a[ny][nx]) >= l && abs(a[y][x] - a[ny][nx]) <= r){
			dfs(ny, nx);
		}
	}
}

int main(){
	cin >> n >> l >> r;
	for(int i = 0; i < n; i++){
		for(int j = 0; j < n; j++){
			cin >> a[i][j];
		}
	}
	while(true){
		bool flag = 0;
		memset(visited, 0, sizeof(visited));
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				if(visited[i][j] == 0){
					v.clear();
					dfs(i, j);
					if(v.size() == 1) continue;
					else{
						flag = 1;
						int sum = 0;
						for(pair<int, int> b : v){
							sum += a[b.first][b.second];
						}
						for(pair<int, int> b : v){
							a[b.first][b.second] = sum / v.size();
						}
					}
				}
			}
		}
		if(!flag) break;
		ret++;
	}
	cout << ret;
}